### DISCUSSION TITLE
Fastest way to export to Arrow for use with Vaex

### DISCUSSION THREAD

**Discussion by maartenbreddels** [duckdb/duckdb#5649] [Q&A]
Hi 👋 

awesome project! I'm the original [Vaex](https://github.com/vaexio/vaex) author, and @JovanVeljanoski pointed me to duckdb, with emphasis on its excellent parquet reading performance. (According to my benchmarks, about 2x as fast as the pyarrow/arrow implementation).
We are interested in using duckdb as a source, which requires exporting the result to arrow, as well as a sink (reading in data from vaex).
I was first exploring using duckdb as a source, and just seeing how fast I can get the data out.

As a reference, I ran this query:
```python
import duckdb
duckdb.query(f"""
    PRAGMA threads=64;
    select sum(fare_amount) from "/data/yellow_taxi_2009_2015_f32.parquet" where fare_amount < 100
""")
```
Runs in about 0.4-0.5 seconds, pretty impressive!

If I want to do this calculation myself, I do something like:
```python
import pyarrow.compute as pc

chunk_size = 1_000_000
con = duckdb.connect()
e = con.execute('''
PRAGMA threads=64;
SELECT fare_amount FROM "/data/yellow_taxi_2009_2015_f32.parquet"
''')
results = e.fetch_record_batch(chunk_size)

total = 1
while True:
    try:
        # Process a single chunk here (just printing as an example)
        def do(batch):
            pass
                
        batch = results.read_next_batch()
        # doesn't matter much if I comment out the next line
        total += pc.sum(batch.columns[0]).as_py()
    except StopIteration:
        break
total
```

This takes 6 seconds, and before exploring further, I was wondering if there is a faster way to extract the data.

**Comment by Alex-Monahan** [comment]
Howdy and welcome! I've followed Vaex for a while now - thanks for reaching out! I like your hybrid numpy/Arrow approach.
Could the difference be that DuckDB benefits from all 64 threads but jumping into Python brings the GIL into play? Is a full Arrow Table just as slow? I know that brings it all into memory...

**Comment by maartenbreddels** [comment]
Thanks :)
I'm not sure it's the GIL. However your question about a full arrow table was a good idea:
```python
import pyarrow.compute as pc

con = duckdb.connect()
e = con.execute('''
PRAGMA threads=64;
SELECT fare_amount FROM "/data/yellow_taxi_2009_2015_f32.parquet"
''')
t = e.arrow()
```
Takes 6.8 seconds, slightly slower, but maybe useful information.

**Comment by maartenbreddels** [comment]
For reference, using arrow:
```python
import pyarrow as pa
import pyarrow.compute as pc
import pyarrow.dataset as ds
taxi = ds.dataset("/data/yellow_taxi_2009_2015_f32.parquet")
total = 0
for batch in taxi.to_batches(columns=["fare_amount"], filter=ds.field("fare_amount") < 100):
    batch.column("fare_amount")
# again, doesn't matter if I include this line or not
#     total += pc.sum(batch.column("fare_amount")).as_py()
print(total)
```

**Comment by yiteng-guo** [comment]
I also played around duckdb in the past few days and tried to figure out what's the fastest way from parquet to arrow. Here're my testing dataset.

```
pq.write_table(
    pa.Table.from_arrays(
        [pa.array(np.random.randn(100_000_000)) for _ in range(5)],
        names=["A", "B", "C", "D", "E"],
    ),
    "/tmp/test-multi-cols.parq",
)
```
The baseline I used here is `pyarrow.parquet.read_table`.

When threads are enabled
```
%time pq.read_table("/tmp/test-multi-cols.parq", use_threads=True);
CPU times: user 723 ms, sys: 3.68 s, total: 4.41 s
Wall time: 762 ms

%time duckdb.sql("SELECT * FROM read_parquet('/tmp/test-multi-cols.parq')").execute();
CPU times: user 1.2 s, sys: 1.46 s, total: 2.66 s
Wall time: 1.76 s

%time duckdb.sql("SELECT * FROM read_parquet('/tmp/test-multi-cols.parq')").arrow(128 * 1024);
CPU times: user 2.61 s, sys: 1.66 s, total: 4.27 s
Wall time: 3.43 s
```

When threads are disabled
```
%time pq.read_table("/tmp/test-multi-cols.parq", use_threads=False);
CPU times: user 767 ms, sys: 3.09 s, total: 3.86 s
Wall time: 2.08 s

duckdb.sql("SET threads TO 1;")

%time duckdb.sql("SELECT * FROM read_parquet('/tmp/test-multi-cols.parq')").execute();
CPU times: user 1.2 s, sys: 1.31 s, total: 2.51 s
Wall time: 2.54 s

%time duckdb.sql("SELECT * FROM read_parquet('/tmp/test-multi-cols.parq')").arrow(128 * 1024);
CPU times: user 2.45 s, sys: 1.54 s, total: 3.99 s
Wall time: 4 s
```

After digging around previous PRs/issues/code, my hypothesis is that `to_arrow` part is sequential. duckdb uses multithreading to fetch parquet data into memory in some duckdb internal representation and then sequentially converts it to arrow table (we also make a copy in this step after #4315 ). 

It's also worth pointing out that we use a python list (i.e. `py::list`) to hold all batches and then call `from_batches` to delegate pyarrow to create the arrow table. Maybe we can instead create a Arrow table in C++ and return that to python? Does this require linking Arrow C++ lib so we don't want to do it?

I understand duckdb isn't the right tool to "bulk load a parquet file" but this slowness could be frustrating when we want to convert a huge output of duckdb to arrow. Definitely let me know if there's any trick I can use to speed this up. It would be much appreciated.

**Reply by Tishj** [reply]
Have you considered fetching a batch reader instead?
This should be streaming, so you don't pay a huge materialization cost upfront

**Reply by yiteng-guo** [reply]
Thanks for the suggestion. I'm able to achieve similar performance as `pq.read_parquet` with this approach when thread is disabled. Code is attached in case anyone find this useful. 
```
b = duckdb.sql("SELECT * FROM read_parquet('/tmp/test-multi-cols.parq')").record_batch(128 * 1024)
bs = b.read_all()
```

However, if I turn on threads, duckdb is still slower than pyarrow.

**Reply by yiteng-guo** [reply]
Also it's worth noting that using batch reader could be slower than using `duckdb.sql(...).arrow()` when it's on NFS (or any slow IO storage) with threading enabled. I guess it's because it's completely sequential since it's steraming. With `arrow`, we pay the huge materialization cost to get the concurrency (on reading).
